lib/jekyll/document.rb

```diff
@@ -242,19 +293,54 @@ def read(opts = {})
       end
     end
 
+    def post_read
+      if DATE_FILENAME_MATCHER =~ relative_path
+        m, cats, date, slug, ext = *relative_path.match(DATE_FILENAME_MATCHER)
+        merge_data!({
+          "slug" => slug,
+          "ext"  => ext
+        })
+        merge_data!({"date" => date}) if data['date'].nil? || data['date'].to_i == site.time.to_i
+        data['title'] ||= slug.split('-').select {|w| w.capitalize! || w }.join(' ')
+      end
+      populate_categories
+      populate_tags
+
+      if generate_excerpt?
+        data['excerpt'] = Jekyll::Excerpt.new(self)
+      end
+    end
+
+    def populate_categories
+      merge_data!({
+        "categories" => (
+          Array(data['categories']) + Utils.pluralized_array_from_hash(data, 'category', 'categories')
+        ).map { |c| c.to_s }.flatten.uniq
+      })
+    end
+
+    def populate_tags
+      merge_data!({
+        "tags" => Utils.pluralized_array_from_hash(data, "tag", "tags").flatten
+      })
+    end
+
     # Create a Liquid-understandable version of this Document.
     #
     # Returns a Hash representing this Document's data.
     def to_liquid
       @to_liquid ||= if data.is_a?(Hash)
-        Utils.deep_merge_hashes data, {
+        Utils.deep_merge_hashes Utils.deep_merge_hashes({
           "output"        => output,
           "content"       => content,
           "relative_path" => relative_path,
           "path"          => relative_path,
           "url"           => url,
-          "collection"    => collection.label
-        }
+          "collection"    => collection.label,
+          "next"          => next_doc,
+          "previous"      => previous_doc,
+          "id"            => id,
```

Comment:
All documents having IDs is going to fix so many terrible workarounds I've created. :smile: